src/dag/write.rs

```diff
@@ -1,55 +1,195 @@
+            Some(h) => self.kvw.put(&head_key, h.as_bytes()).await?,
         }
+
+        self.changed_heads
+            .write()
+            .await
+            .push(HeadChange(hash.map(str::to_string), old_hash));
+
+        Ok(())
     }
 
     pub async fn commit(self) -> Result<()> {
+        self.collect_garbage().await?;
         Ok(self.kvw.commit().await?)
     }
 
     #[allow(dead_code)]
     pub async fn rollback(self) -> Result<()> {
         Ok(self.kvw.rollback().await?)
     }
+
+    async fn set_ref_count(&self, hash: &str, count: u16) -> Result<()> {
+        // Ref count is represented as a u16 stored as 2 bytes using BE.
+        let ref_count_key = Key::ChunkRefCount(hash).to_string();
+        let buf = count.to_be_bytes();
+        if count == 0u16 {
+            self.kvw.del(&ref_count_key).await?;
+        } else {
+            self.kvw.put(&ref_count_key, &buf).await?;
+        }
+        Ok(())
+    }
+
+    async fn get_ref_count(&self, hash: &str) -> Result<u16> {
+        let buf = self.kvw.get(&Key::ChunkRefCount(hash).to_string()).await?;
+        Ok(match buf {
+            None => 0u16,
+            Some(buf) => u16::from_be_bytes([buf[0], buf[1]]),
+        })
+    }
+
+    #[async_recursion(?Send)]
+    async fn change_ref_count(&self, hash: &str, rc: RefCount) -> Result<()> {
+        let old_count = self.get_ref_count(hash).await?;
+        let new_count = (old_count as i32) + (rc as i32);
+
+        if old_count == 0 && rc == RefCount::Inc || old_count == 1 && rc == RefCount::Dec {
+            let meta_key = Key::ChunkMeta(hash).to_string();
+            let buf = self.kvw.get(&meta_key).await?;
+            if let Some(buf) = buf {
+                let meta = meta::get_root_as_meta(&buf);
+                if let Some(refs) = meta.refs() {
+                    for r in refs.iter() {
+                        self.change_ref_count(r, rc).await?;
+                    }
+                }
+            }
+        }
+
+        if new_count == 0 {
+            self.remove_all_related_keys(hash, true).await?;
+        } else {
+            self.set_ref_count(hash, new_count as u16).await?;
+        }
+
+        Ok(())
+    }
+
+    async fn remove_all_related_keys(&self, hash: &str, update_mutated_chunks: bool) -> Result<()> {
+        let data_key = Key::ChunkData(hash).to_string();
+        let meta_key = Key::ChunkMeta(hash).to_string();
+        let ref_count_key = Key::ChunkRefCount(hash).to_string();
+
+        try_join!(
+            self.kvw.del(&data_key),
+            self.kvw.del(&meta_key),
+            self.kvw.del(&ref_count_key),
+        )?;
+
+        // TODO: Getting the write lock could be done in parallel too.
+
+        if update_mutated_chunks {
+            let mut mutated_chunks = self.mutated_chunks.write().await;
+            mutated_chunks.remove(hash);
+        }
+        Ok(())
+    }
+
+    async fn collect_garbage(&self) -> Result<()> {
```

Comment:
[taste] I think it reads better to organize the functions roughly top-down, and Rust doesn't care one way or other.